Context-based solver interface for incremental solvers [depends: 4451, blocks: 4361]

peterschrammel opened the pull request:
Based on #4451 - only review last 4 commits.

Upstreams part of 2LS's push/pop incremental solver interface 

- [X] Each commit message has a non-empty body, explaining why the change was made.
- [X] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [X] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on src/solvers/prop/incremental_solver.h:
```diff
@@ -0,0 +1,64 @@
+  /// Solve the current formula
+  decision_proceduret::resultt dec_solve() override;
+
+  /// Convert a boolean expression \p expr
```
Nit pick: s/boolean/Boolean/

tautschnig commented on src/solvers/prop/incremental_solver.cpp:
```diff
@@ -0,0 +1,60 @@
+incremental_solvert::incremental_solvert(
+  const namespacet &ns,
+  prop_convt &prop_conv)
+  : prop_convt(ns), prop_conv(prop_conv)
```
#4043 makes the useless `namespacet` go away...

peterschrammel replied:
Thanks!!!

tautschnig replied:
Though you'll now have to clean this one up...

tautschnig commented on src/solvers/prop/incremental_solver.h:
```diff
@@ -0,0 +1,64 @@
+
+/// A `prop_convt` that wraps another `prop_convt` and provides
+/// a push/pop-context interface for incremental solving
+class incremental_solvert : public prop_convt
```
I'm aware this is a non-trivial ask, but it would be really nice to have a unit test documenting 1) how this is used and 2) that it works as claimed. Just pushing two or three context and solve a simple Boolean problem.

tautschnig commented on src/solvers/prop/incremental_prop_conv_solver.cpp:
```diff
@@ -0,0 +1,54 @@
+{
+  // We create a new context literal.
+  literalt context_literal = prop_conv.convert(symbol_exprt(
+    "symex::context$" + std::to_string(context_literal_counter++),
```
Make this magic string a static class member?

kroening replied:
"symex" is certainly not happening here.

kroening commented:
The naming here is confusing. The propt interface is already incremental (you could give it the `get_number_of_solver_calls()` method).

What gets added is the push/pop. Call it push_pop_solver? stack_solver?

kroening commented:
Maybe context_solver?

kroening commented:
Also, `set_assumptions` becomes unusable, I suppose? I would then add an implementation with `PRECONDITION(false)`.

kroening commented on src/solvers/prop/context_solver.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+  // unsupported
+  UNREACHABLE;
+}
```
Simply return `false` here!

kroening commented on src/solvers/prop/context_solver.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+  // unsupported
+  UNREACHABLE;
+}
```
Same here

kroening commented:
The hierarchy is not quite right yet.
There should an interface, which extends `decision_proceduret`; it shouldn't prescribe implementation via `propt`.
Then there's a particular kind of implementing this, using `propt`, which will be derived from `prop_convt`.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d726a2b8efa1ab980af3083c8f6e4b736ea68b19).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99902719

tautschnig commented:
@peterschrammel This needs a rebase (and consideration of @kroening's comment).

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: d5eb7811e4b37ec4767d068aca581c34de5a129c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103820299
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 49ac40dd61a353d62dbef4cac3102da76d90356a).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/103823684

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 76774fe621c44578dca237125e626f9f9b4b54cf).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/105672538

martin-cs commented:
Seems like a reasonable thing to want (I note that push/pop seem to be loosing favour in the SMT community in favour of solver-under-assumption).  Shout when this is ready to be reviewed.

peterschrammel committed:
Rename prop_convt to decision_procedure_incrementalt

decision_procedure_incrementalt will then be further split
into support for assumptions and contexts.

peterschrammel committed:
Update decision procedure docs

peterschrammel committed:
Split out decision procedure with solving under assumptions

Makes it explicit which algorithms actually require this feature.

peterschrammel committed:
Decision procedure interface for incremental solving with contexts

Provides push/pop interface

peterschrammel committed:
Split out is_in_conflict solver capability

This is only provided by the prop_conv_solvert-based hierarchy
at the moment and is quite specific to MiniSAT-based solvers.
The functionality itself is used out-of-tree only (2LS).

peterschrammel committed:
Allow getting context-based incremental solver

Puts the type check into a single place

peterschrammel committed:
Implement context-based solver on top of prop_conv_solvert

Also allows using raw assumptions within contexts
(as required by minimization and refinement algorithms).

peterschrammel committed:
Make bv_refinementt compatible with contexts

Must not bypass prop_conv_solvert::set_assumptions by calling
prop directly.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: ac5813189caf20434062b18523e717adf5e909c5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106293396
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 9e504659bbc76aa2b0903c0b4e89bb962b3b9953).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/106306303
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

peterschrammel commented:
Obsolete
